File: libraries/javalib/java/util/Arrays.java

/*
 * Java core library component.
 *
 * Copyright (c) 1999
 *	Archie L. Cobbs.  All rights reserved.
 * Copyright (c) 1999
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 *
 * Author: Archie L. Cobbs <archie@whistle.com>
 */

package java.util;

import java.io.Serializable;

public class Arrays {

  // This class is not instantiable
  private Arrays() {
  }

  // The "default" Comparator
  private static class DefCmp implements Serializable, Comparator {
	public int compare(Object o1, Object o2) {
		return ((Comparable)o1).compareTo(o2);
	}
  }
  static final Comparator DEFAULT_COMPARATOR = new DefCmp();

  public static List asList(Object[] a) {
	return new ArrayList(a);
  }

  public static int binarySearch(byte[] a, byte key) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		if (key == a[index]) {
			return(index);
		} else if (key > a[index]) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(char[] a, char key) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		if (key == a[index]) {
			return(index);
		} else if (key > a[index]) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(double[] a, double key) {
	final long keyBits = Double.doubleToLongBits(key);
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		final long elemBits = Double.doubleToLongBits(a[index]);
		if (keyBits == elemBits) {
			return(index);
		} else if (keyBits > elemBits) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(float[] a, float key) {
	final int keyBits = Float.floatToIntBits(key);
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		final int elemBits = Float.floatToIntBits(a[index]);
		if (keyBits == elemBits) {
			return(index);
		} else if (keyBits > elemBits) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(int[] a, int key) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		if (key == a[index]) {
			return(index);
		} else if (key > a[index]) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(short[] a, short key) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		if (key == a[index]) {
			return(index);
		} else if (key > a[index]) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(long[] a, long key) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		if (key == a[index]) {
			return(index);
		} else if (key > a[index]) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static int binarySearch(Object[] a, Object key) {
  	return binarySearch(a, key, DEFAULT_COMPARATOR);
  }

  public static int binarySearch(Object[] a, Object key, Comparator c) {
	int base = 0;
	for (int lim = a.length; lim != 0; lim >>= 1) {
		final int index = base + (lim >> 1);
		final int diff = c.compare(key, a[index]);
		if (diff == 0) {
			return(index);
		} else if (diff > 0) {
			base = index + 1;
			lim--;
		}
	}
	return ~base;
  }

  public static boolean equals(boolean[] a, boolean[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(byte[] a, byte[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(char[] a, char[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(double[] a, double[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(float[] a, float[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(int[] a, int[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(short[] a, short[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(long[] a, long[] a2) {
	try {
		if (a.length != a2.length) {
			return false;
		}
		for (int i = a.length; i-- > 0; ) {
			if (a[i] != a2[i]) {
				return false;
			}
		}
		return true;
	} catch (NullPointerException _) {
		return (a == a2);	// ie, they are both null
	}
  }

  public static boolean equals(Object[] a, Object[] a2) {
		if (a == a2) {
			return true;
		}

		if (a == null || a2 == null) {      
			return false; // One is null and not the other.
		}

		if (a.length != a2.length) {
			return false;
		}

		for (int i = a.length; i-- > 0; ) {
			if (a[i] == null) {
				if (a2[i] != null) {
					return false;
				}
			} else if (!a[i].equals(a2[i])) {
				return false;
			}
		}

		return true;
  }

  public static void fill(boolean[] a, boolean val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(boolean[] a, int fromIndex, int toIndex, boolean val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(byte[] a, byte val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(byte[] a, int fromIndex, int toIndex, byte val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(char[] a, char val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(char[] a, int fromIndex, int toIndex, char val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(double[] a, double val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(double[] a, int fromIndex, int toIndex, double val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(float[] a, float val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(float[] a, int fromIndex, int toIndex, float val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(int[] a, int val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(int[] a, int fromIndex, int toIndex, int val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(short[] a, short val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(short[] a, int fromIndex, int toIndex, short val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(long[] a, long val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(long[] a, int fromIndex, int toIndex, long val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void fill(Object[] a, Object val) {
	fill(a, 0, a.length, val);
  }

  public static void fill(Object[] a, int fromIndex, int toIndex, Object val) {
	for (int i = toIndex; i-- > fromIndex; ) {
		a[i] = val;
	}
  }

  public static void sort(byte[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(byte[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortByte(a, fromIndex, toIndex);
  }

  public static void sort(char[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(char[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortChar(a, fromIndex, toIndex);
  }

  public static void sort(double[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(double[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortDouble(a, fromIndex, toIndex);
  }

  public static void sort(float[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(float[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortFloat(a, fromIndex, toIndex);
  }

  public static void sort(int[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(int[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortInt(a, fromIndex, toIndex);
  }

  public static void sort(short[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(short[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortShort(a, fromIndex, toIndex);
  }

  public static void sort(long[] a) {
	sort(a, 0, a.length);
  }

  public static void sort(long[] a, int fromIndex, int toIndex) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortLong(a, fromIndex, toIndex);
  }

  public static void sort(Object[] a) {
	sort(a, 0, a.length, DEFAULT_COMPARATOR);
  }

  public static void sort(Object[] a, Comparator c) {
	sort(a, 0, a.length, c);
  }

  public static void sort(Object[] a, int fromIndex, int toIndex) {
	sort(a, fromIndex, toIndex, DEFAULT_COMPARATOR);
  }

  public static void sort(Object[] a, int fromIndex, int toIndex, Comparator c) {
	if (fromIndex < 0 || toIndex > a.length || fromIndex > toIndex) {
		throw new ArrayIndexOutOfBoundsException();
	}
	sortObject(a, fromIndex, toIndex, c);
  }

  private static native void sortByte(byte[] a, int fromIndex, int toIndex);
  private static native void sortChar(char[] a, int fromIndex, int toIndex);
  private static native void sortDouble(double[] a, int fromIndex, int toIndex);
  private static native void sortFloat(float[] a, int fromIndex, int toIndex);
  private static native void sortInt(int[] a, int fromIndex, int toIndex);
  private static native void sortShort(short[] a, int fromIndex, int toIndex);
  private static native void sortLong(long[] a, int fromIndex, int toIndex);
  private static native void sortObject(Object[] a, int fromIndex, int toIndex, Comparator c);

}

File: libraries/clib/native/Arrays.c

/*
 * java.util.Arrays.c
 *
 * Copyright (c) 1999
 *	Archie L. Cobbs. All rights reserved.
 * Copyright (c) 1999
 *	Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution 
 * of this file. 
 *
 * Author: Archie L. Cobbs <archie@whistle.com>
 */

#include "config.h"
#include "config-std.h"
#include "native.h"
#include "gtypes.h"
#include "fp.h"
#include "support.h"
#include "java_util_Comparator.h"
#include "java_util_Arrays.h"

/*
 * Sorting functions
 *
 * We rely on libc routines to do the sorting because they are already
 * highly optimized and debugged.
 *
 * Note that these routines assume the bounds checking has already been
 * done in Java.
 */

static int
cmpByte(const void *p1, const void *p2)
{
	return (int)*((const jbyte *) p1) - (int)*((const jbyte *) p2);
}

static int
cmpChar(const void *p1, const void *p2)
{
	return (int)*((const jchar *) p1) - (int)*((const jchar *) p2);
}

static int
cmpDouble(const void *p1, const void *p2)
{
	const jlong bits1 = doubleToLong(*((const jdouble *) p1));
	const jlong bits2 = doubleToLong(*((const jdouble *) p2));

	return (bits1 == bits2) ? 0 : (bits1 < bits2) ? -1 : 1;
}

static int
cmpFloat(const void *p1, const void *p2)
{
	const jint bits1 = floatToInt(*((const jfloat *) p1));
	const jint bits2 = floatToInt(*((const jfloat *) p2));

	return (bits1 == bits2) ? 0 : (bits1 < bits2) ? -1 : 1;
}

static int
cmpInt(const void *p1, const void *p2)
{
	const jint int1 = *((const jint *) p1);
	const jint int2 = *((const jint *) p2);

	return (int1 == int2) ? 0 : (int1 < int2) ? -1 : 1;
}

static int
cmpShort(const void *p1, const void *p2)
{
	return (int)*((const jshort *) p1) - (int)*((const jshort *) p2);
}

static int
cmpLong(const void *p1, const void *p2)
{
	const jlong long1 = *((const jlong *) p1);
	const jlong long2 = *((const jlong *) p2);

	return (long1 == long2) ? 0 : (long1 < long2) ? -1 : 1;
}

void
java_util_Arrays_sortByte(HArrayOfByte *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpByte);
}

void
java_util_Arrays_sortChar(HArrayOfChar *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpChar);
}

void
java_util_Arrays_sortDouble(HArrayOfDouble *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpDouble);
}

void
java_util_Arrays_sortFloat(HArrayOfFloat *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpFloat);
}

void
java_util_Arrays_sortInt(HArrayOfInt *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpInt);
}

void
java_util_Arrays_sortShort(HArrayOfShort *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpShort);
}

void
java_util_Arrays_sortLong(HArrayOfLong *a, jint fromIndex, jint toIndex)
{
	qsort(&unhand_array(a)->body[fromIndex], (size_t)(toIndex - fromIndex),
		sizeof(*unhand_array(a)->body), cmpLong);
}

/*
 * Sorting object arrays with a Comparator
 *
 * This is less than elegant. We want to use the libc mergesort(), but we
 * get a race condition if we store the Comparator in a static variable.
 * So we allocate a "shadow" array and sort that instead, then copy the
 * object pointers back into the original array.
 */

/* Shadow array element */
struct objcmpinfo {
	struct Hjava_lang_Object *obj;		/* the Object */
	struct Hjava_util_Comparator *cmp;	/* the Comparator  */
};

static int
cmpObject(const void *p1, const void *p2)
{
	const struct objcmpinfo	*const o1 = (const struct objcmpinfo *) p1;
	const struct objcmpinfo	*const o2 = (const struct objcmpinfo *) p2;
	jvalue rtn;

	do_execute_java_method(&rtn, o1->cmp, "compare",
		"(Ljava/lang/Object;Ljava/lang/Object;)I",
		NULL, false, o1->obj, o2->obj);
	return rtn.i;
}

void
java_util_Arrays_sortObject(HArrayOfObject *a, jint fromIndex, jint toIndex, struct Hjava_util_Comparator *c)
{
	int slen = toIndex - fromIndex;
	unsigned int len;
	struct objcmpinfo *ilist;
	errorInfo info;
	unsigned int k;

	if (slen <= 1) {
		return;
	}
	len = slen;

	/* Prepare shadow array */
	ilist = KMALLOC(len * sizeof(*ilist));
	if (ilist == NULL) {
		goto nomem;
	}
	for (k = 0; k < len; k++) {
		ilist[k].obj = unhand_array(a)->body[fromIndex + k];
		ilist[k].cmp = c;
	}

	/* Do the sort */
#ifdef HAVE_MERGESORT
	if (mergesort(ilist, len, sizeof(*ilist), cmpObject) < 0) {
		KFREE(ilist);
		goto nomem;
	}
#else
	qsort(ilist, len, sizeof(*ilist), cmpObject);
#endif

	/* Copy back sorted results */
	for (k = 0; k < len; k++) {
		unhand_array(a)->body[fromIndex + k] = ilist[k].obj;
	}
	KFREE(ilist);
	return;

nomem:
	postOutOfMemory(&info);
	throwError(&info);
}

